azurerm/resource_arm_ad_application.go

```diff
@@ -0,0 +1,325 @@
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MinItems: 1,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"reply_urls": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MinItems: 1,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"available_to_other_tenants": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Computed: true,
+			},
+
+			"oauth2_allow_implicit_flow": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Computed: true,
+			},
+
+			"key_credential": keyCredentialsSchema(),
+
+			"password_credential": passwordCredentialsSchema(),
+
+			"app_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"object_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func customizeDiffAd(diff *schema.ResourceDiff, v interface{}) error {
+
+	if err := customizeDiffKeyCredential(diff, v); err != nil {
+		return err
+	}
+
+	if err := customizeDiffPasswordCredential(diff, v); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceArmAdApplicationCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).applicationsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("display_name").(string)
+	multitenant := d.Get("available_to_other_tenants").(bool)
+
+	properties := graphrbac.ApplicationCreateParameters{
+		DisplayName:             &name,
+		Homepage:                expandAzureRmAdApplicationHomepage(d, name),
+		IdentifierUris:          expandAzureRmAdApplicationIdentifierUris(d, name),
+		ReplyUrls:               expandAzureRmAdApplicationReplyUrls(d, name),
+		AvailableToOtherTenants: utils.Bool(multitenant),
+	}
+
+	if v, ok := d.GetOk("oauth2_allow_implicit_flow"); ok {
+		properties.Oauth2AllowImplicitFlow = utils.Bool(v.(bool))
+	}
+
+	if _, ok := d.GetOk("key_credential"); ok {
+		keyCreds, err := expandAzureRmKeyCredentials(d, nil)
+		if err != nil {
+			return err
+		}
+		properties.KeyCredentials = keyCreds
+	}
+
+	if _, ok := d.GetOk("password_credential"); ok {
+		passCreds, err := expandAzureRmPasswordCredentials(d, nil)
+		if err != nil {
+			return err
+		}
+		properties.PasswordCredentials = passCreds
+	}
+
+	app, err := client.Create(ctx, properties)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*app.ObjectID)
+
+	return resourceArmAdApplicationRead(d, meta)
+}
+
+func resourceArmAdApplicationRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).applicationsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	resp, err := client.Get(ctx, d.Id())
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[DEBUG] Application ID %q was not found - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error loading Application %q: %+v", d.Id(), err)
+	}
+
+	d.Set("display_name", resp.DisplayName)
```

Comment:
can we access all of these properties on `resp.[XX]Properties`? it's how they're returned from the API - these fields aren't necessarily guaranteed to be populated. This also means we can do a nil-check e.g.

```
if props := resp.[XX]Properties; props != nil {
  d.Set("name", resp.DisplayName)
  # ...
}
```